src/Stratis.Bitcoin.Features.Consensus/CoinViews/CachedCoinView.cs

```diff
@@ -427,7 +427,8 @@ public void SaveChanges(IList<UnspentOutputs> unspentOutputs, IEnumerable<TxOut[
                             for (int i = 0; i < unspent.Outputs.Length; i++)
                             {
                                 // Only push to rewind data index UTXOs that are spent.
-                                if (unspent.Outputs[i] == null && clone.Outputs[i] != null)
+                                // Checks against array length are needed in case problem like 3965 on VSTS appears.
+                                if ((i < unspent.Outputs.Length) && (i < clone.Outputs.Length) && (unspent.Outputs[i] == null) && (clone.Outputs[i] != null))
```

Comment:
first `if` check is redundant because it's already enforced by the `for`, you can remove it ( `(i < unspent.Outputs.Length) `)

Reply:
How is this even possible?
`unspent` and `clone` are the same copy of each other?

Reply:
this is exactly what I was talking right now with Ivan
anyway the cloned object is the same but then we override Outputs with this
https://github.com/stratisproject/StratisBitcoinFullNode/blob/994f5052c18e540d4d9c82306508d047c9cc0760/src/Stratis.Bitcoin.Features.Consensus/CoinViews/CachedCoinView.cs#L413

The question is: is this a valid scenario?

Reply:
(cacheItem.UnspentOutputs is the result of FetchCoins )

Reply:
yes it is actually see my other comment